Quote SELECT identifiers only if autoquoting is enabled.

ADmad opened the pull request.

ADmad committed:
Quote SELECT identifiers only if autoquoting is enabled.

ADmad commented:
Postgres is playing spoilt sport :cry:. I don't have it locally to check if there's some easy fix that can be done to avoid the failures.

ADmad commented:
So apparently Posgres auto converts all unquoted identifier to lower cased which is what's causing the failures.

We could make a new `PostgresCompiler` (similar to existing `SqliteCompiler`) and retain the quoting only for Postgres.

ADmad commented:
If creating a `PostgresCompiler` is acceptable I'll make required changes and reopen this against `4.next`. Even though query compilers are marked internal I don't want to much around with them in `4.0`

markstory commented:
@ADmad I think having a postgres specific compiler is fair. We have compilers for Sqlite and Sqlserver to work around their specific requirements.

ADmad committed:
Always quote SELECT identifier for Postgres.

Postgres converts unquoted identifiers to lower cased which causes problems
with CamelCased aliases.

ADmad commented:
@markstory I eventually decided to add `PostgresCompiler` in this PR itself for `4.0`. I think it's best to make this quoting related change early in 4.x life cycle. Waiting till 4.1 increases the chance of causing problems to users and quite a few would have already adopted Cake 4.

othercorey commented on src/Database/QueryCompiler.php:
```diff
@@ -85,6 +85,13 @@ class QueryCompiler
+     *
+     * @var bool
+     */
+    protected $_quotedSelectIdentifiers = false;
```
This is specifically for quoting aliases right, not column identifiers.

ADmad replied:
Yes, should I rename the property to better reflect that?

othercorey replied:
I think so.

markstory commented on src/Database/QueryCompiler.php:
```diff
@@ -85,6 +85,13 @@ class QueryCompiler
+     *
+     * @var bool
+     */
+    protected $_quotedSelectAliases = false;
```
I'm half tempted to not have protected property overrides and instead duplicate `_buildSelectPart` into the Postgres compiler so that both the postgres and non-postgres versions have no additional branches. Our test suite is thorough enough that we should be able to keep changes aligned between the two implementations.

ADmad replied:
I too had contemplated the same thing :)

I will update accordingly.

ADmad replied:
@markstory Duplicating `_buildSelectPart()` will only avoid a `if` check in `PostgresCompiler`. The `if` check will still be required in `QueryCompiler` if auto quoting is enabled.

markstory replied:
Well that doesn't seem worth it.
